File: Universal_Gates/test_gen.py
#!/usr/bin/env python
# coding: utf-8

# In[3]:


import math

# Function to convert decimal to binary with fixed length
def dec_to_bin(num, length):
    return format(num, f'0{length}b')

# Function to perform NAND operation
def nand_2(A, B):
    return int(not (A and B))

# Function to perform AND operation
def and_2(A, B):
    return int(A and B)

# Function to perform OR operation
def or_2(A, B):
    return int(A or B)

# Function to perform XOR operation
def xor_2(A, B):
    return int(A ^ B)

# Function to perform XNOR operation
def xnor_2(A, B):
    return int(not (A ^ B))

# Function to perform half-adder operation
def half_adder(A, B):
    sum = xor_2(A, B)
    carry = and_2(A, B)
    return sum, carry

# Main function to generate the trace file
def generate_tracefile():
    output_file = "TRACEFILE.txt"
    input_len = 2  # Input vector length for 2-input gates
    output_len = 1 # Output vector length for 2-input gates
    test_len = 2 ** input_len  # Total number of test vectors (exhaustive)
    
    with open(output_file, "w") as optr:
        for i in range(test_len):
            input_bin = dec_to_bin(i, input_len)  # Get input in binary form

            # Separate the inputs
            A = int(input_bin[0])
            B = int(input_bin[1])

            # Perform the gate operations
            nand_out = nand_2(A, B)
            and_out = and_2(A, B)
            or_out = or_2(A, B)
            xor_out = xor_2(A, B)
            xnor_out = xnor_2(A, B)
            
            # Perform half-adder operation
            sum, carry = half_adder(A, B)

            # Combine all outputs in the required order
            outputs = [
                sum,        # Sum bit (MSB)
                carry,      # Carry bit (second MSB)
                xnor_out,   # XNOR output (third MSB)
                xor_out,    # XOR output
                or_out,     # OR output
                and_out,    # AND output
                nand_out    # NAND output (LSB)
            ]

            # Create the output binary string
            output_bin = ''.join(dec_to_bin(out, output_len) for out in outputs)

            # Mask bits (equal to number of output bits)
            mask = "1" * len(outputs)

            # Write input, output, and mask to the trace file
            optr.write(f'{input_bin} {output_bin} {mask}\n')

# Run the main function
if __name__ == "__main__":
    generate_tracefile()



# Open/create TRACEFILE.txt in write mode
f = open("TRACEFILE.txt", "w")

# Number of inputs
input_len = 2

# Number of outputs
output_len = 1

# Number of test vectors (2^input_len)
test_len = 2 ** input_len

# Loop through all possible input vectors (from 0 to test_len-1)
for input_vec in range(test_len):
    # Convert input to binary (2-bit) format
    input_str = "{:02b}".format(input_vec)
    
    # Extract individual input bits
    a = (input_vec >> 1) & 1  # Most significant bit
    b = input_vec & 1         # Least significant bit
    
    # Calculate outputs for gates
    nand_output = int(not (a & b))
    and_output = a & b
    or_output = a | b
    xor_output = a ^ b
    xnor_output = int(not (a ^ b))
    
    # Calculate outputs for half-adder
    sum_output = xor_output
    carry_output = and_output
    
    # Combine all outputs in the required order
    # sum as MSB, carry as second MSB, XNOR as third MSB, and so on.
    outputs = [
        sum_output,   # Sum bit (MSB)
        carry_output, # Carry bit (second MSB)
        xnor_output,  # XNOR output (third MSB)
        xor_output,   # XOR output
        or_output,    # OR output
        and_output,   # AND output
        nand_output   # NAND output (LSB)
    ]
    
    # Create the output binary string
    output_str = ''.join(str(out) for out in outputs)

    # Mask bits (equal to number of output bits)
    mask = "1" * len(outputs)

    # Write input, output, and mask to the trace file
    f.write(f"{a}{b} {output_str} {mask}\n")

# Close the file
f.close()

File: Universal_Gates/test_gen.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

// File pointer as global variable
FILE *optr;

// Function prototypes
void dec_to_bin(long long num, int len); // Function to convert decimal to binary and print in TRACEFILE.txt
int bin_to_dec(long long num);           // Function to convert binary to decimal for inputs

// Function definition for decimal to binary conversion
void dec_to_bin(long long num, int len)
{
    // Array to store binary values (initial length set to 20, change it according to your design)
    int bin[20];
    int rem, i, j;
    i = 0;

    // Simple binary conversion, pen paper logic
    while (num != 0)
    {
        rem = num % 2;
        num /= 2;
        bin[i] = rem;
        ++i;
    }

    // Print binary values with MSB first (padded with extra 0s)
    for (j = len - 1; j >= 0; j--)
    {
        if (j < i)
            fprintf(optr, "%d", bin[j]);
        else
            fprintf(optr, "0");
    }
}

// Function definition for binary to decimal conversion
int bin_to_dec(long long num)
{
    int i, dec, rem;
    dec = 0;
    i = 0;
    // Simple digit extraction logic
    while (num != 0)
    {
        rem = num % 10;
        num /= 10;
        dec += rem * pow(2, i);
        ++i;
    }
    return dec;
}

// Logic gate functions
void nand_2(int A, int B, int *Y) {
    *Y = !(A && B);
}

void and_2(int A, int B, int *Y) {
    *Y = A && B;
}

void or_2(int A, int B, int *Y) {
    *Y = A || B;
}

void xor_2(int A, int B, int *Y) {
    *Y = A ^ B;
}

void xnor_2(int A, int B, int *Y) {
    *Y = !(A ^ B);
}

// Half-adder functions
void half_adder(int A, int B, int *sum, int *carry) {
    *sum = A ^ B;
    *carry = A & B;
}

int main()
{
    char output_file[20];
    int input_len, output_len, test_len;
    int i, A, B, nand_out, and_out, or_out, xor_out, xnor_out, sum, carry;

    // Name the tracefile
    strcpy(output_file, "TRACEFILE.txt");
    // Open TRACEFILE.txt in write mode
    optr = fopen(output_file, "w");

    // Input and output vector length for 2-input gates
    input_len = 2;
    output_len = 1;
    // Total number of test vectors (exhaustive) = 2^(number of input bits)
    test_len = pow(2, input_len);

    // Run loop for all possible combinations of input for each gate
    for (i = 0; i < test_len; i++) {
        // Separate the inputs (binary form)
        A = (i / 2) % 2;    // Extract bit A
        B = i % 2;          // Extract bit B

        // Perform the gate operations
        nand_2(A, B, &nand_out);
        and_2(A, B, &and_out);
        or_2(A, B, &or_out);
        xor_2(A, B, &xor_out);
        xnor_2(A, B, &xnor_out);
        
        // Perform half-adder operation
        half_adder(A, B, &sum, &carry);

        // Print input in "TRACEFILE.txt" (in binary)
        dec_to_bin(i, input_len);

        // Single space
        fprintf(optr, " ");

        // Print outputs for all gates and half-adder in "TRACEFILE.txt" (in binary)
        dec_to_bin(sum, output_len);        // Sum bit (MSB)
        dec_to_bin(carry, output_len);      // Carry bit (second MSB)
        dec_to_bin(xnor_out, output_len);   // XNOR output (third MSB)
        dec_to_bin(xor_out, output_len);    // XOR output
        dec_to_bin(or_out, output_len);     // OR output
        dec_to_bin(and_out, output_len);    // AND output
        dec_to_bin(nand_out, output_len);   // NAND output (LSB)

        // Single space
        fprintf(optr, " ");

        // Mask bits (equal to number of output bits)
        for (int j = 0; j < 7; j++) {
            fprintf(optr, "1");
        }

        // Change line
        fprintf(optr, "\n");
    }

    // Close file (TRACEFILE.txt)
    fclose(optr);
    return 0;
}
